### PR TITLE
Initial IMAP connector implementation

### PR DESCRIPTION
**Pull request by foundingnimo**
Tested with Office365

Won't work with Gmail

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,51 @@
+{
+  "name": "reshuffle-imap-connector",
+  "version": "0.0.1",
+  "description": "An Reshuffle IMAP connector",
```
Comment (lagache) [inline package.json]:
```suggestion
  "description": "A Reshuffle IMAP connector",
```

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  on(options: IMAPConnectorEventOptions, eventId: string): EventConfiguration {
+    if (!eventId) {
+      eventId = `IMAP/${options.name}/${this.id}`
+    }
+    const event = new EventConfiguration(eventId, this, options)
+    this.eventConfigurations[event.id] = event
+    return event
+  }
+
+  onRemoveEvent(event: EventConfiguration) {
```
Comment (lagache) [inline src/index.ts]: not sure if we need this onRemoveEvent, it doesn't seem to be used in Reshuffle

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  private imapReady() {
+    this.imap.openBox(this.mailbox, false, (error, mailbox) => {
+      if (error) throw error
+
+      this.imap.on('mail', (newMessageNumber: any) => {
+        this.handleNewMessages()
+      })
+    })
+  }
+
```
Comment (lagache) [inline src/index.ts]: BOS provides a search action fro IMAP, don't we want it here?

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    const event = new EventConfiguration(eventId, this, options)
+    this.eventConfigurations[event.id] = event
+    return event
+  }
+
+  onRemoveEvent(event: EventConfiguration) {
+    delete this.eventConfigurations[event.id]
+  }
+
+  private imapReady() {
```
Comment (lagache) [inline src/index.ts]: I don't get what this method does, maybe the method name doesn't reflect what it is for
also missing return type

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+export interface IMAPConnectorEventOptions {
+  name: string
+}
+
+class IMAPConnector extends BaseConnector<IMAPConnectorConfigOptions, IMAPConnectorEventOptions> {
+  private imap: ImapClient
+  private parser = simpleParser
+  private readonly mailbox: string
+
+  constructor(options: IMAPConnectorConfigOptions, id: string) {
```
Comment (lagache) [inline src/index.ts]: I don't think options and id are mandatory

File: src/index.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
+import { Reshuffle, BaseConnector, EventConfiguration } from 'reshuffle-base-connector'
+import ImapClient from 'imap'
+import { simpleParser } from 'mailparser'
+
+export interface IMAPConnectorConfigOptions {
+  host: string
```
Comment (lagache) [inline src/index.ts]: Are they all mandatory?
What is the markSeen option? shouldn't be an action?

File: src/index.ts (hunk 3 above)
Comment (foundingnimo) [reply src/index.ts]: Yes - 2nd PR

File: src/index.ts (hunk 4 above)
Comment (foundingnimo) [reply src/index.ts]: Good catch :) Old code.

File: package.json (hunk 1 above)
Comment (foundingnimo) [reply package.json]: cool thanks!

File: src/index.ts (hunk 2 above)
Comment (foundingnimo) [reply src/index.ts]: This is a wider Reshuffle thing..

File: src/index.ts (hunk 5 above)
Comment (foundingnimo) [reply src/index.ts]: options are in IMAP. id not

File: src/index.ts (hunk 6 above)
Comment (foundingnimo) [reply src/index.ts]: `markSeen` is part of IMAP.
Changed a few to be non-mandatory

**Pull request merged**